### PR TITLE
ARROW-8958: [FlightRPC][Python] implement DoExchange

### PR DESCRIPTION
**Pull request by lidavidm**

### CODE REVIEW

**Comment by lidavidm** [comment]
It seems https://issues.apache.org/jira/browse/ARROW-7689 has come back. The [gRPC issue](https://github.com/grpc/grpc/issues/20311) has been closed for lack of activity, I suppose this means we should just disable any TLS tests on MacOS.

**Comment by lidavidm** [comment]
Or actually, it was supposedly fixed in https://github.com/grpc/grpc/pull/21804, so we should try bumping to gRPC >= 1.28.

**Comment by lidavidm** [comment]
Attempting to bump gRPC in https://github.com/apache/arrow/pull/7298.

File: python/pyarrow/_flight.pyx [hunk 1]
```diff
@@ -844,8 +856,27 @@ cdef class FlightStreamReader(MetadataRecordBatchReader):
             (<CFlightStreamReader*> self.reader.get()).Cancel()
 
 
-cdef class FlightStreamWriter(_CRecordBatchWriter):
-    """A RecordBatchWriter that also allows writing application metadata."""
+cdef class MetadataRecordBatchWriter(_CRecordBatchWriter):
+    """A RecordBatchWriter that also allows writing application metadata.
+
+    This class is a context manager; on exit, close() will be called.
```
Comment (pitrou) [inline python/pyarrow/_flight.pyx:862]: Is `close` a method in the base class? If so, how about defining `__enter__` and `__exit__` in the base class instead?

**Review by pitrou — APPROVED** [review]
LGTM, just one comment

File: python/pyarrow/_flight.pyx (hunk 1 above)
Comment (lidavidm) [reply python/pyarrow/_flight.pyx:862]: Turns out this was already in the base class, so I removed the definition here.

Thanks for the review!

**Pull request closed without merging**